keps/sig-api-machinery/2839-in-use-protection/README.md

```diff
@@ -0,0 +1,682 @@
+
+#### Namespace Deletion
+
+It may also not be clear how [namespace deletion](https://kubernetes.io/docs/tasks/administer-cluster/namespaces/#deleting-a-namespace) behaves with lien.
+Therefore, this section describes the behavior.
+
+In summary, lien will block the namespace deletion, but would not block deletion of any resources in the namespace individually.
+
+If a lien is set to a namespace, only a deletion request to the namespace is blocked.
+Users can still request to delete each object in the namespace directly, but they can't request to delete it through the namespace deletion.
+
+If a lien is set to one of objects in a namespace, only a deletion request to the object is blocked.
+Users can still request to delete its namespace or other objects in the namespace.
+When deletion of the namespace A is requested, the deletion isn't completed until the object with lien is deleted.
+
+#### Block Adding Additional `Liens` while Deleting
+
+`Liens` shouldn't be added after `DeletionTimestamp` is non-nil, which means pre-deletion processes for finalizers are being handled.
+Adding `Liens` while deleting itself won't do any harm for the cluster, because the resource will be deleted without any additional deletion requests that will be blocked by `Liens`.
+However, users may think that the successful addition of the `Liens` means that the resource won't be deleted until the `Liens` are deleted, which isn't true.
+To avoid such a misunderstanding from users, a request for adding additional `Liens` to a resource with non-nil `DeletionTimestamp` should be blocked by api-server.
+
+#### Format of `Liens`
+
+As a result of discussions, format of `Liens` is decided as a slice of strings and it needs to be used per controller(per user) basis.
+The main reasons are scalability and better aligned with `Finalizers`.
+
+- Per controller(Per user) `Liens`
+  - Pros:
+    - Aligned with `Finalizers`
+    - Good scalability (Size of the field and number of API calls required)
+  - Cons:
+    - Bad UX (Users won't know which other objects are actually using the object)
+    - Complex consuming controllers' implementation (Each controller needs to track the using objects)
+- Per reason(Per object) `Liens`
+  - Pros:
+    - Good UX (Users can know which other objects are actually using the object)
+    - Simple consuming controllers' implementation (Lien tracks the using objects)
+  - Cons:
+    - Not aligned with `Finalizers`
+    - Bad scalability (Size of the field and number of API calls required)
+
+#### Race of removing/adding liens
+
+A cluster-admin would be racing to update the object to remove the lien and delete the object before a namespace editor is able to place the lien back. As a result, a namespace admin or editor can create an object that a cluster-admin or namespace lifecycle controller cannot delete, which shouldn't be allowed.
+
+To mitigate the risk, a new API option in `DeleteOptions` to force delete a resource with liens will be added.
```

Comment:
At the heart of this feature is the fact that it breaks all existing deletion flows the moment a user adds a lien and the client tool is not aware of it.

That’s a very big change - tools will not work, libraries will not work, etc.

Given our commitment to backward compatibility, I’m really not comfortable with breaking all client deletions.  Instead, I believe liens must be precondition for delete, not a “on by default behavior” (ie Precondition - no leins from specfic tools).

If there is more justification for “break api by default” it needs to be more prominent here (or i’m blind) but I don’t think what is here meets the bar for such a massive disruption.

Reply:
> Given our commitment to backward compatibility, I’m really not comfortable with breaking all client deletions. Instead, I believe liens must be precondition for delete, not a “on by default behavior” (ie Precondition - no leins from specfic tools).

Could you explain a bit more about your concern and idea to solve it?

Are you worrying about adding Lien itself because it breaks the existing behavior or worrying about adding an option to ignore Lien because it breaks the behavior of Lien?

Reply:
@smarterclayton 

I think at least for the accidental deletion use case, any thorough mechanism has the property that it breaks workflows deleting the protected object. So you are effectively arguing against thorough mechanisms, e.g. that deletion protection should only affect kubectl or opt-in clients. I think this is not right (e.g. a motivator is saving an object when you forget to turn on Terraform's deletion protection). Since your argument is not against this particular mechanism, I think you have to argue against the use case.

So: for the accidental deletion protection use case, preventing deletion flows is the point and I think that's good, I expect it to be somewhat rare.

Elsewhere in the KEP it's decreed that namespace deletion will always force, but garbage collection should not force. I do not think there's a reasonable design that lacks the "ignore liens" permission. We could also have an "add/update liens" permission, but since at least some users would have that permission, at best that just reduces the number of protected objects, and therefore I don't think it's necessary if important actors can force.

I agree that if you do something stupid like protect all of your pods, the pod-gc controller won't clean them up and your cluster will fill up and stop working. IMO misuses are possible all over the system and while it's arguably our responsibility to make misuse inconvenient or at least not easy, it's not our responsibility to make it impossible; that standard rules out too many useful mechanisms.

Now, in terms of automated application of liens, I think you have a great point, and I think we should debate that separately at each proposed use. This KEP just makes the mechanism possible; we definitely would have to be responsible when using it in an automated fashion. E.g. I am actually not sure that the proposed application for keeping secrets around when a pod uses them is a good one, it sounds risky. But I'm fairly sure that the PVC/PV relationship could benefit from liens.

Reply:
> breaks all existing deletion flows the moment a user adds a lien and the client tool is not aware of it.

Which deletion flow breaks in particular?  Delete requests can fail for any number of reasons.  Since clients already have to handle deletion failures and since the semantics of what happens on a successful delete call does not change, the change will be limited to non-humans (humans could read the message and add --no-really--force-this-time).

The intended application of this approach is to prevent that non-human from taking an action that shouldn't happen (otherwise there wouldn't be a lien).

We're now constrained down to non-human clients that, when operating correctly, should override the expressed desire to avoid deleting a resource.  That is a fairly constrained set.

Reply:
@lavalamp i can buy the argument that that “use of leins” is what is breaking compatibility, in which case i’d be fine if this KEP articulated that and specified that automated leins are what take the burden of proving api compatibility, and thus a kep introducing them has to pass that bar.

Reply:
Question for breakage: is there some way that leins can be overridden for automated tools *without* needing to change the code of the tools?  I'm thinking administrator action, but done outside what may be a third-party tool. If the answer is yes, then there's no compatibility problem.

If the answer is no, it becomes like deleting a table in Postgres 9, where you literally had to kick all users out of the database before you could drop a popular table (or its primary key index).  That's not a good workflow; sometimes you do need to delete things even though they are in use.

Reply:
The first thing that comes to mind is to add an "auto-override-liens" permission which could be given to particular automated tools.

Before adding that or something similar to the design, I personally would like to hear from more users first that yes, they want some automated tooling to ignore deletion protection. Because I expected people to mostly want protection *from* the automated tooling.

The next step on this is "I want object A to be deletable by controller X even if A is protected. However I want object B to be super-protected so that not even controller X can delete it. However controller Y should still be able to delete object B in that case. I also have object C which should not be deletable by X or Y, but controller Z may delete it. ..." It is easy to model this as an integer instead of as a boolean (e.g. put a number on both the lien and the permission) but I really hope it's not necessary.

Reply:
@lavalamp I'm thinking of cases that go like this:

1. Deploy #3721 failed, could not delete object in use
2. Admin: Can we stop object from being used?  No.
3. Admin: override to push through Deploy #3721 and delete the object in use

So not a blanket deletion authorization for a particular role (although that would be good too, for admins) but a one-time exception for specific operations.  The key is being able to execute on that *without* changing the source code of the CI/CD system.

Reply:
Also, if you keep it binary, and for specific operations, then the whole "graph problem" doesn't occur.  Either the operation has authorization to ignore liens, or it doesn't.

Reply:
> Question for breakage: is there some way that leins can be overridden for automated tools without needing to change the code of the tools? I'm thinking administrator action, but done outside what may be a third-party tool. If the answer is yes, then there's no compatibility problem.

As an advisory feature, any users who can update an object can force delete the object by (1) patching the object to remove all the liens on the object, and then (2) deleting the object, in most cases. Some users may think that this extra step to remove liens break the existing tool, but at least they can continue using their tools without changing the tools' codes or behaviors. I hope the answer to your question can be "yes" by this behavior.

Note that the force delete option here is for admins to avoid a special case where a kind of controllers, maybe not a human, watches the object and keep setting liens on it. Admins may eventually win the race to remove the liens before deletion, but it would be better to provide a way to avoid such situations. Also, allowing only certain users to specify this option is to avoid users or tools from always specifying this option, which makes this entire feature useless. 

In addition, for consuming controllers to avoid such races, they can be implemented to only add liens in certain conditions, like an object has a certain annotation, as described [here](https://github.com/kubernetes/enhancements/pull/2640/files#diff-7c2b9b5fce93fb956c6589eaefc44491414766ee66821cd4ddd4e4294449ab4bR186-R189). Such mechanism can be either opt-in or opt-out, but should be able to be changed later after creating objects. By using this kind of features on the consuming controllers side, users or tools will be able to delete objects without force delete and race.